static-check/src/types/circuits/circuit.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    types::circuits::{CircuitFunctionType, CircuitVariableType},
+    Attribute,
+    FunctionType,
+    SymbolTable,
+    Type,
+    TypeError,
+};
+use leo_typed::{Circuit, CircuitMember, Identifier, InputValue, Parameter, Span};
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// Stores circuit definition details.
+///
+/// This type should be added to the circuit symbol table for a resolved syntax tree.
+/// This is a user-defined type.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
```

Comment:
if the `identifier` is supposed to be unique, I'd definitely introduce hand-written `PartialEq` and `Hash` impls in order to work only on the `identifier` as opposed to checking/hashing the collections too